src/Storages/StorageS3.cpp

```diff
@@ -263,6 +264,9 @@ class StorageS3Source::DisclosedGlobIterator::Impl : WithContext
             outcome_future = listObjectsAsync();
         }
 
+        if (result_batch.empty())
+            throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "Can not match any files using prefix {}", request.GetPrefix());
```

Comment:
but we call this method `fillInternalBufferAssumeLocked` more than once, here we can have `!is_finished`, it means that we will return to this method and collect the next batch of files later and we will do that as long as !is_finished, so on the last iteration `result_batch` will be empty but it does not mean that s3 wildcard did not match any file.

Reply:
More over we cannot throw regardless, it must be under setting and by default `false`.

Reply:
Also to fix current implementation need to check the following: this is the first iteration of getting result of list objects request && is_finished && result_batch.empty()

Reply:
as @xiedeyantu pointed out in DM, the last iteration will not be empty as it will not come into this method as it will check `is_finished` flag and not call it, so no need for the check in my last comment